File: Session4/206. Reverse Linked List.java
class Solution {
    ListNode ans;
    public ListNode reverseList(ListNode head) {
        return reverseList_rec(head);
       // return ans;
    }


  public ListNode reverseList_rec(ListNode head) {
        if(head==null || head.next==null )
        {
            return head;
        }
        ListNode temp=reverseList(head.next);
        head.next.next=head;
        head.next=null;
        return temp;
    }
}


class Solution {
    public ListNode reverseList(ListNode head) {
       ListNode prev=null;
       ListNode curr=head;
       ListNode next=null;
       while(curr!=null)
       {
           next=curr.next;
           curr.next=prev;
           prev=curr;
           curr=next;
       }
        return prev;
    }
}

File: Session4/206. Reverse Linked List.py
class Solution:
        
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        return self.helper(head)
    
    def helper(self,head):
        if head is None or head.next is None:
            return head
        temp=self.helper(head.next)
        head.next.next=head
        head.next=None
        return temp

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return 
        
        prev=None
        
        while head.next:
            temp=head.next
            head.next=prev
            prev=head
            head=temp
        head.next=prev
        
        return head
            